README.rst
========================
Pyramid Method Rewriting
========================

The ``pyramid_methodrewrite`` package is a pyramid plugin that adds a
"tween" that rewrites the active HTTP method (as exposed via
``request.method``) to the value provided in a HTTP header or
query-string. The main reason for this is to allow clients to fake the
use of HTTP methods beyond GET and POST (such as PUT and DELETE) that
are behind non-compliant proxies, browsers, or other deficient
software that limit which HTTP methods can be sent.

It does this by looking for an ``X-HTTP-Method-Override`` HTTP header
or a ``_method`` parameter in either the query-string or POST data,
and if the value is of the known set of HTTP methods, will override
the active request's ``.method`` attribute. Both the header name and
parameter name are configurable. The header value, if specified, takes
precedence over the parameter value.

For example, the request:

.. code-block:: text

  GET /path/to/resource?_method=OPTIONS HTTP/1.1

Becomes equivalent (from the application's point of view) to:

.. code-block:: text

  OPTIONS /path/to/resource HTTP/1.1


Project
=======

* Homepage: https://github.com/cadithealth/pyramid_methodrewrite
* Bugs: https://github.com/cadithealth/pyramid_methodrewrite/issues


Installation
============


.. code-block:: bash

  $ pip install pyramid-methodrewrite


Usage
=====

Enable the tween either in your INI file via:

.. code-block:: ini

  pyramid.includes = pyramid_methodrewrite

or in code in your package's application initialization via:

.. code-block:: python

  def main(global_config, **settings):
    # ...
    config.include('pyramid_methodrewrite')
    # ...


Configuration
=============

The following configuration settings can be set in your application's
``main`` section:

* `methodrewrite.enabled`: a boolean flag to control whether or not
  method rewriting should be enabled. Default: ``True``.

* `methodrewrite.on`: a list of incoming HTTP methods that will cause
  this tween to check the request parameters and potentially make the
  change. Defaults to ``(GET, POST)``.

* `methodrewrite.header`: the name of the header that controls the
  requested method. Defaults to ``X-HTTP-Method-Override``. If
  present, the header override takes precedence over the parameter
  override.

* `methodrewrite.param`: the name of the parameter that controls the
  requested method. Defaults to ``_method``.

* `methodrewrite.to`: list of acceptable HTTP methods that a request
  can be rewritten to. If set to empty, no restrictions will be made
  on mapping to acceptable or known methods. Defaults to
  ``pyramid_methodrewrite.HTTP_METHODS``.

Note that all method names are always converted to upper case.


CHANGELOG.rst
=========
ChangeLog
=========


v0.2.3
======

* Removed `distribute` dependency


v0.2.2
======

* Added support for HTTP header based override (in addition to
  query-string parameter based), by default using the
  ``X-HTTP-Method-Override`` header name


v0.2.1
======

* First tagged release


pyramid_methodrewrite/test.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# auth: Philip J Grabner <grabner@cadit.com>
# date: 2013/04/08
# copy: (C) Copyright 2013 Cadit Inc., see LICENSE.txt
#------------------------------------------------------------------------------

import unittest
from pyramid import testing
from pyramid.request import Request
from pyramid_methodrewrite import factory, includeme

#------------------------------------------------------------------------------
class TestRewriteFactory(unittest.TestCase):

  def setUp(self):
    self.config = testing.setUp()
    self.registry = self.config.registry

  def test_register(self):
    handler = factory(None, self.registry)
    self.assertEqual(handler.__name__, 'methodrewrite_tween')

#------------------------------------------------------------------------------
class TestRewriteTween(unittest.TestCase):

  def setupRequest(self, url='/', headers=None):
    request = Request.blank(url, headers=headers)
    self.request = request
    self.config = testing.setUp(request=request)
    self.registry = self.config.registry
    self.registry.settings = {}
    request.registry = self.registry

  def setupTween(self):
    # includeme(self.config)
    self.tween = factory(self.handler, self.registry)

  def handler(self, request):
    return request.method + ' ' + request.path_qs

  def test_direct(self):
    self.setupRequest('/path')
    self.setupTween()
    self.assertEqual(self.tween(self.request), 'GET /path')

  def test_rewrite_put_param(self):
    self.setupRequest('/path?_method=put')
    self.setupTween()
    self.assertEqual(self.tween(self.request), 'PUT /path?_method=put')

  def test_rewrite_put_header(self):
    self.setupRequest('/path', {'x-http-method-override': 'PUT'})
    self.setupTween()
    self.assertEqual(self.tween(self.request), 'PUT /path')

  def test_limit_to_no_put_param(self):
    self.setupRequest('/path?_method=put')
    self.registry.settings['methodrewrite.to'] = 'DELETE SEARCH'
    self.setupTween()
    self.assertEqual(self.tween(self.request), 'GET /path?_method=put')

  def test_limit_to_no_put_header(self):
    self.setupRequest('/path', {'x-http-method-override': 'PUT'})
    self.registry.settings['methodrewrite.to'] = 'DELETE SEARCH'
    self.setupTween()
    self.assertEqual(self.tween(self.request), 'GET /path')

  def test_limit_to_with_delete_param(self):
    self.setupRequest('/path?_method=delete')
    self.registry.settings['methodrewrite.to'] = 'DELETE SEARCH'
    self.setupTween()
    self.assertEqual(self.tween(self.request), 'DELETE /path?_method=delete')

  def test_limit_to_with_delete_header(self):
    self.setupRequest('/path', {'X-HTTP-METHOD-OVERRIDE': 'delete'})
    self.registry.settings['methodrewrite.to'] = 'DELETE SEARCH'
    self.setupTween()
    self.assertEqual(self.tween(self.request), 'DELETE /path')

  def test_change_name_param(self):
    self.setupRequest('/path?_method=delete')
    self.registry.settings['methodrewrite.param'] = 'rewrite'
    self.setupTween()
    self.assertEqual(self.tween(self.request), 'GET /path?_method=delete')
    self.setupRequest('/path?rewrite=delete')
    self.registry.settings['methodrewrite.param'] = 'rewrite'
    self.setupTween()
    self.assertEqual(self.tween(self.request), 'DELETE /path?rewrite=delete')

  def test_change_name_header(self):
    self.setupRequest('/path', {'x-http-method-override': 'DELETE'})
    self.registry.settings['methodrewrite.header'] = 'x-methodrewrite'
    self.setupTween()
    self.assertEqual(self.tween(self.request), 'GET /path')
    self.setupRequest('/path', {'x-methodrewrite': 'DELETE'})
    self.registry.settings['methodrewrite.header'] = 'x-methodrewrite'
    self.setupTween()
    self.assertEqual(self.tween(self.request), 'DELETE /path')

#------------------------------------------------------------------------------
# end of $Id$
#------------------------------------------------------------------------------


pyramid_methodrewrite/__init__.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# lib:  pyramid_methodrewrite
# auth: Philip J Grabner <grabner@cadit.com>
# date: 2013/04/08
# copy: (C) Copyright 2013 Cadit Inc., see LICENSE.txt
#------------------------------------------------------------------------------

from pyramid.settings import aslist, asbool

#------------------------------------------------------------------------------

DEFAULT_PARAM_NAME   = '_method' 
DEFAULT_HEADER_NAME  = 'X-HTTP-Method-Override'

HTTP_METHODS = (

  # shamelessly scrubbed from:
  #   http://annevankesteren.nl/2007/10/http-methods
  # todo: need to do some real research and bump this.

  # RFC 2616 (HTTP 1.1):
  'OPTIONS',
  'GET',
  'HEAD',
  'POST',
  'PUT',
  'DELETE',
  'TRACE',
  'CONNECT',

  # RFC 2518 (WebDAV):
  'PROPFIND',
  'PROPPATCH',
  'MKCOL',
  'COPY',
  'MOVE',
  'LOCK',
  'UNLOCK',

  # RFC 3253 (WebDAV versioning):
  'VERSION-CONTROL',
  'REPORT',
  'CHECKOUT',
  'CHECKIN',
  'UNCHECKOUT',
  'MKWORKSPACE',
  'UPDATE',
  'LABEL',
  'MERGE',
  'BASELINE-CONTROL',
  'MKACTIVITY',

  # RFC 3648 (WebDAV collections):
  'ORDERPATCH',

  # RFC 3744 (WebDAV access control):
  'ACL',

  # draft-dusseault-http-patch:
  'PATCH',

  # draft-reschke-webdav-search:
  'SEARCH',

)

#------------------------------------------------------------------------------
def factory(handler, registry):
  get   = registry.settings.get
  on    = [e.upper() for e in aslist(get('methodrewrite.on', 'GET POST'))]
  to    = [e.upper() for e in aslist(get('methodrewrite.to', ' '.join(HTTP_METHODS)))]
  pname = get('methodrewrite.param', DEFAULT_PARAM_NAME)
  hname = get('methodrewrite.header', DEFAULT_HEADER_NAME)
  def methodrewrite_tween(request):
    if request.method.upper() not in on:
      return handler(request)
    if hname in request.headers:
      meth = request.headers.get(hname, '').upper()
      if meth and ( not to or meth in to ):
        request.method = meth
        return handler(request)
    if pname in request.params:
      meth = request.params.get(pname, '').upper()
      if meth and ( not to or meth in to ):
        request.method = meth
        return handler(request)
    return handler(request)
  return methodrewrite_tween

#------------------------------------------------------------------------------
def includeme(config):
  '''
  Adds a pyramid :term:`tween` to `config` that converts GET or POST
  requests that have a '_method' GET or POST parameter set to a known
  HTTP method to have that method explicitly. Note that no other
  request rewriting is done to account for changes in interface.

  The following configuration settings can be set:

  * `methodrewrite.enabled`: a boolean flag to control whether or not
    method rewriting should be enabled. Default: ``True``.

  * `methodrewrite.on`: a list of incoming HTTP methods that will
    cause this tween to check the request parameters and potentially
    make the change. Defaults to ``(GET, POST)``.

  * `methodrewrite.header`: the name of the header that controls the
    requested method. Defaults to ``_method``. If present, the header
    override takes precedence over the parameter override.

  * `methodrewrite.param`: the name of the parameter that controls the
    requested method. Defaults to ``_method``.

  * `methodrewrite.to`: list of acceptable HTTP methods that a request
    can be rewritten to. If set to empty, no restrictions will be made
    on mapping to acceptable or known methods. Defaults to
    ``pyramid_methodrewrite.HTTP_METHODS``.

  Note that all methods are always converted to upper case.
  '''
  if asbool(config.registry.settings.get('methodrewrite.enabled', 'true')):
    config.add_tween('pyramid_methodrewrite.factory')

#------------------------------------------------------------------------------
# end of $Id$
#------------------------------------------------------------------------------
